### PR TITLE
repo_utils: support github pull request branches

### PR DESCRIPTION
**Pull request by kshtsk**
This adds possibility to checkout github pr providing special purpose ref spec (like refs/pull/X/merge and refs/pull/X/head) as branch parameter to teuthology, for example, "teuthology-suite -t".

### CODE REVIEW

**Comment by kshtsk** [comment]
@susebot run deploy

**Comment by kshtsk** [comment]
@susebot run deploy

**Comment by kshtsk** [comment]
@susebot run deploy

**Comment by kshtsk** [comment]
@susebot run deploy

File: teuthology/repo_utils.py [hunk 1]
```diff
@@ -218,12 +269,16 @@ def reset_repo(repo_url, dest_path, branch):
                       GitError for other errors
     """
     validate_branch(branch)
-    log.info('Resetting repo at %s to branch %s', dest_path, branch)
+    if '/' in branch:
+        reset_branch = lsstrip(remote_ref_from_ref(branch), 'refs/remotes/')
+    else:
+        reset_branch = 'origin/' + branch
```
Comment (smithfarm) [inline teuthology/repo_utils.py]: Why change `'origin/%s' % branch` to `'origin/' + branch` here?

File: teuthology/repo_utils.py [hunk 2]
```diff
@@ -255,7 +310,8 @@ def fetch_repo(url, branch, bootstrap=None, lock=True):
     src_base_path = config.src_base_path
     if not os.path.exists(src_base_path):
         os.mkdir(src_base_path)
-    dirname = '%s_%s' % (url_to_dirname(url), branch)
+    branch_dir = ref_to_dirname(branch)
+    dirname = '%s_%s' % (url_to_dirname(url), branch_dir)
```
Comment (smithfarm) [inline teuthology/repo_utils.py]: if `+` (concat) operator is preferred (to get as many Exceptions as possible?), then it should be used here, too

And it would be nice to have some rationale in the commit message explaining why this change (from `%s` to `+`) is being made.

File: teuthology/repo_utils.py [hunk 3]
```diff
@@ -138,6 +142,53 @@ def clone_repo(repo_url, dest_path, branch, shallow=True):
...
+    branch_name = local_branch_from_ref(ref)
+    remote_ref = remote_ref_from_ref(ref)
+    misc.sh('git init %s' % dest_path)
+    misc.sh('git remote add origin %s' % repo_url, cwd=dest_path)
+    #misc.sh('git fetch --depth 1 origin %s' % fetch_refspec(ref),
+    #                                                        cwd=dest_path)
+    fetch_branch(dest_path, ref)
+    misc.sh('git checkout -b %s %s' % (branch_name, remote_ref),
+                                                            cwd=dest_path)
+
```
Comment (smithfarm) [inline teuthology/repo_utils.py]: one more newline here

Comment (kshtsk) [reply teuthology/repo_utils.py]: there is a 2 new lines not three, see cwd=dest_path at the end.

File: teuthology/repo_utils.py (hunk 1 above)
Comment (kshtsk) [reply teuthology/repo_utils.py]: It was changed with no special intend.

File: teuthology/repo_utils.py (hunk 2 above)
Comment (kshtsk) [reply teuthology/repo_utils.py]: Not the case :-)

**Comment by kshtsk** [comment]
@susebot run deploy

**Comment by djgalloway** [comment]
Interesting!  How does this work without all the packaging CI (Jenkins/shaman/chacra)?

**Comment by kshtsk** [comment]
> Interesting! How does this work without all the packaging CI (Jenkins/shaman/chacra)?

This is supposed to be useful for --teuthology-branch, and I guess for --suite-branch too. However, the --ceph could theoretically work with `--newest`, but I'm not sure.

But again, this is mainly for testing teuthology PRs.

**Comment by djgalloway** [comment]
> > Interesting! How does this work without all the packaging CI (Jenkins/shaman/chacra)?
> 
> This is supposed to be useful for `--teuthology-branch`, and I guess for `--suite-branch` too. However, the --ceph could theoretically work with `--newest`, but I'm not sure.
> 
> But again, this is mainly for testing teuthology PRs.

Ah, okay.  Gotcha.  I misunderstood the purpose.
